### PR TITLE
feat: Deprecate `BasePageNumberPaginator` and `BaseOffsetPaginator` in favor of `PageNumberPaginator` and `OffsetPaginator`, respectively

### PR DESCRIPTION
**Pull request by edgarrmondragon**
These were made "base" classes with the expectation that users would override them, but https://github.com/meltano/sdk/pull/1918 made that redundant.

## Related

- https://github.com/meltano/sdk/pull/1918
- https://github.com/meltano/sdk/pull/2989
- https://github.com/meltano/sdk/pull/3483

## Summary by Sourcery

Deprecate the base paginator classes in favor of concrete PageNumberPaginator and OffsetPaginator and update usages, tests, and docs accordingly.

New Features:
- Introduce PageNumberPaginator and OffsetPaginator as the primary paginator implementations for page-number and offset-based pagination.
- Add a shared singer_sdk_deprecated decorator helper preconfigured with the SDK's deprecation warning type.

Bug Fixes:
- Tighten paginator-related tests by asserting concrete types and cleaning up header handling in header-based pagination tests.

Enhancements:
- Mark BasePageNumberPaginator and BaseOffsetPaginator as deprecated wrappers around the new paginator classes, exposing page_size as a property on OffsetPaginator.
- Update REST stream examples, cookiecutter templates, and dummyjson tap client to use the new paginator classes and type hints, including explicit override annotations and typed Context usage.

Documentation:
- Revise pagination documentation and cookiecutter guidance to recommend PageNumberPaginator and OffsetPaginator, and add reference stubs for their API documentation.

Tests:
- Extend and adjust pagination tests to cover deprecation warnings for the base paginator classes and to use the new PageNumberPaginator and OffsetPaginator implementations.

## Summary by Sourcery

Deprecate legacy base pagination classes in favor of concrete page-number and offset paginator implementations and update usages, tests, and docs accordingly.

New Features:
- Introduce PageNumberPaginator and OffsetPaginator as the primary page-number and offset-based paginator implementations.
- Add a shared singer_sdk_deprecated helper preconfigured with the SDK's deprecation warning type for consistent deprecation messaging.

Bug Fixes:
- Tighten header-based and link-based pagination tests by asserting concrete types and correcting header handling edge cases.

Enhancements:
- Mark BasePageNumberPaginator and BaseOffsetPaginator as deprecated wrappers around the new paginator classes, exposing page_size as a property on OffsetPaginator.
- Update REST stream examples, cookiecutter templates, and the dummyjson tap client to use the new paginator classes, override annotations, and typed Context parameters.

Documentation:
- Revise pagination guides and cookiecutter documentation to recommend PageNumberPaginator and OffsetPaginator and add API reference stubs for their classes.

Tests:
- Extend pagination tests to cover deprecation warnings for the base paginator classes and to exercise the new concrete paginator implementations.

### CODE REVIEW

File: singer_sdk/pagination.py [hunk 1]
```diff
@@ -405,6 +422,17 @@ def get_next(self, response: requests.Response) -> int | None:
         return self._value + self._page_size
 
 
+@singer_sdk_deprecated(
+    "BaseOffsetPaginator is deprecated and will be removed in a future version. "
+    "Use OffsetPaginator instead."
+)
+class BaseOffsetPaginator(OffsetPaginator):
+    """DEPRECATED   .
```
Comment (edgarrmondragon) [inline singer_sdk/pagination.py]:
```suggestion
    """DEPRECATED.
```
